browser/chat/chat_tab_helper.cc

```diff
@@ -0,0 +1,207 @@
+  chat_history_.push_back(turn);
+}
+
+void ChatTabHelper::GetArticleSummaryString(
+    OnArticleSummaryCallback on_article_summary_callback) {
+  if (!article_summary_.empty()) {
+    VLOG(1) << __func__ << " Summary coming from cache: " << article_summary_;
+
+    std::move(on_article_summary_callback).Run(article_summary_, true);
+    return;
+  }
+
+  ui::AXTreeID tree_id = web_contents()->GetPrimaryMainFrame()->GetAXTreeID();
+  content::RenderFrameHost* rfh =
+      content::RenderFrameHost::FromAXTreeID(tree_id);
+
+  rfh->GetMainFrame()->RequestAXTreeSnapshot(
+      base::BindOnce(&ChatTabHelper::OnSnapshotFinished, base::Unretained(this),
+                     std::move(on_article_summary_callback)),
+      ui::AXMode::kWebContents, false, 5000, {});
+}
+
+void ChatTabHelper::SetArticleSummaryString(const std::u16string text) {
+  article_summary_ = text;
+}
+
+void ChatTabHelper::AddObserver(Observer* observer) {
+  observers_.AddObserver(observer);
+}
+
+void ChatTabHelper::RemoveObserver(Observer* observer) {
+  observers_.RemoveObserver(observer);
+}
+
+void ChatTabHelper::OnSnapshotFinished(
+    OnArticleSummaryCallback on_article_summary_callback,
+    const ui::AXTreeUpdate& snapshot) {
+  ui::AXTree tree;
+  if (!tree.Unserialize(snapshot)) {
+    return;
+  }
+
+  // Start AX distillation process
+  // Don't copy the tree, as it can be expensive.
+  DistillViaAlgorithm(std::move(on_article_summary_callback), std::move(tree));
+}
+
+void ChatTabHelper::DistillViaAlgorithm(
+    OnArticleSummaryCallback&& on_article_summary_callback,
+    const ui::AXTree&& tree) {
+  std::vector<const ui::AXNode*> content_root_nodes;
+  std::vector<const ui::AXNode*> content_nodes;
+  GetContentRootNodes(tree.root(), &content_root_nodes);
+
+  for (const ui::AXNode* content_root_node : content_root_nodes) {
+    AddContentNodesToVector(content_root_node, &content_nodes);
+  }
+
+  std::vector<std::u16string> text_node_contents;
+  for (const ui::AXNode* content_node : content_nodes) {
+    AddTextNodesToVector(content_node, &text_node_contents);
+  }
+
+  // TODO(nullhook): The assumption here is that 9300 chars equate to
+  // approximately 2k tokens, which is a rough estimate. A proper tokenizer is
+  // needed for accurate measurement.
+  auto contents_text =
+      base::JoinString(text_node_contents, u" ").substr(0, 9300);
+
+  VLOG(1) << __func__
+          << " Number of chars in content text = " << contents_text.length()
+          << "\n";
+
+  std::move(on_article_summary_callback).Run(std::move(contents_text), false);
+}
+
+void ChatTabHelper::PrimaryPageChanged(content::Page& page) {
+  CleanUp();
+
+  for (auto& obs : observers_) {
+    obs.OnPageChanged();
+  }
+}
+
+void ChatTabHelper::WebContentsDestroyed() {
+  CleanUp();
+}
+
+void ChatTabHelper::CleanUp() {
+  chat_history_.clear();
+  article_summary_.clear();
+}
+
+WEB_CONTENTS_USER_DATA_KEY_IMPL(ChatTabHelper);
```

Comment:
I'm wondering manipulating string here is on browser's UI thread. Then, can that cause a hang when processing a gigantic article?

Reply:
I can't tell which line of code you're referring to